sdk/core/Azure.Core/src/Shared/GeoRedundantReadFallbackPolicy.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core.Pipeline;
+
+namespace Azure.Core.Shared
+{
+    /// <summary>
+    ///
+    /// </summary>
+    internal class GeoRedundantReadFallbackPolicy : HttpPipelineSynchronousPolicy
+    {
+        private readonly string[] _readFallbackHosts;
+        private int? _fallbackIndex;
+        private readonly object _syncLock = new();
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="readFallbackHosts"></param>
+        public GeoRedundantReadFallbackPolicy(string[] readFallbackHosts)
+        {
+            _readFallbackHosts = readFallbackHosts;
+        }
+
+        public static void SetHostAffinity(HttpMessage message, bool hostAffinity)
+        {
+            message.SetProperty(typeof(HostAffinityKey), hostAffinity);
+        }
+
+        private static bool GetHostAffinity(HttpMessage message)
+        {
+            return message.TryGetProperty(typeof(HostAffinityKey), out object? hostAffinity) && hostAffinity is true;
+        }
+
+        private static void SetPrimaryHost(HttpMessage message)
+        {
+            message.SetProperty(typeof(PrimaryHostKey), message.Request.Uri.Host!);
+        }
+
+        private static string GetPrimaryHost(HttpMessage message)
+        {
+            message.TryGetProperty(typeof(PrimaryHostKey), out object? primaryHost);
+            return (string)primaryHost!;
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="message"></param>
+        public override void OnSendingRequest(HttpMessage message)
+        {
+            // TODO implement policy for write hosts as well (or include in this policy)
+            if (message.Request.Method != RequestMethod.Get && message.Request.Method != RequestMethod.Head)
+                return;
+
+            if (message.HasResponse || GetHostAffinity(message))
+                return;
+
+            if (message.ProcessingContext.RetryNumber == 0)
+            {
+                SetPrimaryHost(message);
+                // avoid locking for common path - there is a chance that the index would be incremented here by another thread but we prioritize
+                // throughput over correctness in this case
+                if (_fallbackIndex == null)
+                    return;
+            }
+
+            // must be under lock so that this can be synchronized across client calls - each call should leverage the fallback information
+            // from all other calls
+            lock (_syncLock)
```

Comment:
Do we need to lock, or we could use Interlocked?

Reply:
I think we need some form of lock to handle Increment index + update host to use that index. Using interlocked can allow us to increment index atomically, but it might get incremented again before updating the host.